### PR TITLE
Update: Allow to omit semi for one-line blocks (fixes #4385)

### LINKED ISSUE #4385: Add an option to 'semi' for omitting semicolon before brace on one-line ( JSHint's 'lastsemic' )
**When does this rule warn? Please describe and show example code:**
Current 'semi' has two options 'always' or 'never'. But I want more smart semi constraint.

```
// Allow this
var reduced = list.reduce(function (r, i) { return r + i });

// But disallow this
var reduced = list.reduce(function (r, i) {
    return r + i 
});

```

Actually, this is smart exception of 'always'.

**Is this rule preventing an error or is it stylistic?**
This is stylistic issue.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**
JSHint has an option like this. ref. http://jshint.com/docs/options/#lastsemic
I'm considering to use eslint instead of jshint. People who use this option may want same behavior of this.

**Are you willing to create the rule yourself?**
PR: #4386

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28154261-add-an-option-to-semi-for-omitting-semicolon-before-brace-on-one-line-jshint-s-lastsemic?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by alberto**

### CODE REVIEW

File: docs/rules/semi.md [hunk 1]
```diff
@@ -133,6 +141,54 @@ var name = "ESLint"
 })()
 ```
 
+#### "always-except-oneline"
```
Comment (nzakas) [inline docs/rules/semi.md]: Not 100% sure about the option name. @eslint/eslint-team thoughts?

**Comment by alberto** [comment]
Would adding `block` at the end make it more clear? I also had doubts about `oneline` vs `single-line`

File: docs/rules/semi.md (hunk 1 above)
Comment (IanVS) [inline docs/rules/semi.md]: I think this could be confusing for people who may conflate "except" with "accept".  It reads too much like a sentence.

This actually seems like it should be a separate option, like:

```json
{
    "semi": [2, "always", "except-oneline"]
}
```

Not sure about `oneline` vs. `single-line`, but I think I like `oneline` better.

**Comment by nzakas** [comment]
I think the aspect that is most important is the concept of the last statement in the block. The JSHint option is `lastsemic`, and while that's a bit too terse, I think the word "last" is an important one.

**Comment by alberto** [comment]
So, are you proposing `except-last-semi`?
What about `except-last-in-oneliner`?

**Comment by nzakas** [comment]
I've tried to stay away from options with more than two dashes, as I think those get to be a bit hard on the eyes. To have multiple words, we probably need to do something like:

```
semi: [2, "always", {
    omitLastInOneLineBlock: true
}]
```

And have `omitLastInOneLineBlock` only apply when using `"always"` (that can be configured as part of the schema). A bit wordy, but at least it's clear.

File: docs/rules/semi.md (hunk 1 above)
Comment (alberto) [inline docs/rules/semi.md]: @IanVS It is a separate option

**Comment by alberto** [comment]
What about:
```
semi: [2, "always", {
    exceptions: ["lastInOneLineBlock"]
}]
```

**Comment by nzakas** [comment]
`exceptions` is ambiguous (we've been trying to not use that in rules going forward). We also don't tend to use strings for toggling things on off, only for matches (like ID matches or AST matches), so I think using properties with true/false is the right approach here.

**Comment by alberto** [comment]
Ok, it's good to know that about exceptions.

**Comment by alberto** [comment]
Do you know of a similar rule? I can't get the schema syntax right. I'm trying to enforce the object only for the `always` case, but I don't know how to define that. I tried the following, but I don't know how to define that the second alternative should have two properties, it shouldn't really be an array.
```
module.exports.schema = [
    {
        "oneOf": [
            {
                "enum": ["always", "never"]
            },
            {
                type: "array",
                items: [
                    {
                        "enum": ["always"]
                    },
                    {
                        "type": "object",
                        "properties": {
                            "omitLastInOneLineBlock": {"type": "boolean"}
                        },
                        "additionalProperties": false
                    }                    
                ]
            }
        ]
    }
];
```

**Comment by btmills** [comment]
@alberto If I understand correctly, you're going for something similar to `curly`. I usually use http://jsonschemalint.com/draft4/ when working on schemas. Try this:

```json
{
    "anyOf": [
        {
            "type": "array",
            "items": [
                {
                    "enum": ["never"]
                }
            ],
            "minItems": 0,
            "maxItems": 1
        },
        {
            "type": "array",
            "items": [
                {
                    "enum": ["always"]
                },
                {
                    "type": "object",
                    "properties": {
                        "omitLastInOneLineBlock": {"type": "boolean"}
                    },
                    "additionalProperties": false
                }                    
            ],
            "minItems": 0,
            "maxItems": 2
        }
    ]
}
```

**Comment by alberto** [comment]
Thanks @btmills! That did it

File: docs/rules/semi.md (hunk 1 above)
Comment (nzakas) [inline docs/rules/semi.md]: @alberto can you update the docs too?

**Comment by gyandeeps** [comment]
@alberto Correct me if I am wrong but the current docs doesnt jibe well with the implementation (option name wise)

**Comment by alberto** [comment]
@gyandeeps you are right. I totally forgot to update the docs with the new structure

**Comment by alberto** [comment]
Docs updated

**Comment by gyandeeps** [comment]
@alberto I am sorry but i am still confused. where is the implementation of `always-except-oneline` option?

**Comment by alberto** [comment]
@gyandeeps Sorry, I forgot to remove the old doc. `always-except-oneline` was changed to `{ "omitLastInOneLineBlock": true}`. Thanks for your help.

**Comment by nzakas** [comment]
Lgtm

**Comment by tandrewnichols** [comment]
Is there any rule or option to make omitLastInOneLineBlock _optional_ rather than _required_? Meaning: we have

```
semi: [2, "always", {
    omitLastInOneLineBlock: true
}]
```

but that _requires_ that you have _no_ semi-colon in a single-line function. Is there any laxer rule/option that says "Require a semi-colon everywhere and just ignore single line functions whether or not they have a semi-colon"?
